### ISSUE TITLE
Union should be supertype of its members

### ISSUE THREAD

**Issue by layday** [microsoft/pyright#472]
**Describe the bug**
This was brought up recently in the mypy tracker and I believe pyright handles it incorrectly.  Take the following example:

```py
# pyright: strict

from typing import Dict, Union

foo: Dict[Union[int, str], str] = {}
bar: Dict[str, str] = {}

foo = bar
```

The assignment should be invalid because dictionaries are invariant and a union is, nominally, a [supertype of its members](https://www.python.org/dev/peps/pep-0484/#id29).  Python dictionaries are mutable and names hold references to their values.  This means that both `foo` and `bar` refer to the same object.  If you were to add a key `0` to 'foo', that would be reflected on 'bar' and would be incompatible with its type annotation.

**To Reproduce**
Type check the snippet above.

**Expected behavior**
That you should not be able to assign 'bar' to 'foo'.

**Screenshots or Code**
See above.

**VS Code extension or command-line**
extension, 1.1.17

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release of Pyright.

**Comment by erictraut** [comment]
This is fixed in 1.1.18, which I just published.
